src/internal/log/level.go

```diff
@@ -0,0 +1,89 @@
+package log
+
+import (
+	"sync/atomic"
+	"time"
+
+	"go.uber.org/zap/zapcore"
+)
+
+// resettableLevel is a log level (zapcore.LevelEnabler) that can be changed at runtime either
+// forever (SetLevel) or for a set duration (SetLevelFor).  See also: zap.AtomicLevel.
+//
+// A resettableLevel must be initialized before use with NewResettableLevelAt.  A resettableLevel's
+// methods take a value receiver; the contained pointers are the actual value of the object (and do
+// not change as the level is changed).
+type resettableLevel struct {
+	orig, cur *atomic.Int32
+	reset     *atomic.Pointer[time.Timer]
+}
+
+var _ zapcore.LevelEnabler = resettableLevel{}
+
+// NewResettableLevelAt creates a new resettable level set to the provided level, and set to revert
+// to that level if changed with SetLevelFor.
+func NewResettableLevelAt(l zapcore.Level) resettableLevel {
+	rl := resettableLevel{
+		orig:  new(atomic.Int32),
+		cur:   new(atomic.Int32),
+		reset: new(atomic.Pointer[time.Timer]),
+	}
+	rl.orig.Store(int32(l))
+	rl.cur.Store(int32(l))
+	return rl
+}
+
+// Level returns the current level.
+func (rl resettableLevel) Level() zapcore.Level {
```

Comment:
2 pointers copied by value. I don't think there's anything incorrect about this.

A `*resettableLevel` with non-pointer values in the struct would probably have fewer de-references and better memory locality.  Not a blocker.

Reply:
We talked about this a little in Slack.  I kind of got "talked into" this pattern by zap.AtomicLevel, which takes non-pointer receivers and has a pointer to the *atomic.Int32 that stores the level.